sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/Triggers.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.windowing;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import java.io.Serializable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.common.runner_api.v1.RunnerApi;
+import org.apache.beam.sdk.transforms.windowing.AfterWatermark.AfterWatermarkEarlyAndLate;
+import org.apache.beam.sdk.transforms.windowing.Never.NeverTrigger;
+import org.apache.beam.sdk.transforms.windowing.Trigger.OnceTrigger;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** Utilities for working with {@link Triggers Triggers}. */
+@Experimental(Experimental.Kind.TRIGGER)
+public class Triggers implements Serializable {
+
+  @VisibleForTesting static final ProtoConverter CONVERTER = new ProtoConverter();
+
+  public static RunnerApi.Trigger triggerToProto(Trigger trigger) {
+    return CONVERTER.convertTrigger(trigger);
+  }
+
+  @VisibleForTesting
+  static class ProtoConverter {
+
+    public RunnerApi.Trigger convertTrigger(Trigger trigger) {
+      Method evaluationMethod = getEvaluationMethod(trigger.getClass());
+      return tryConvert(evaluationMethod, trigger);
+    }
+
+    private RunnerApi.Trigger tryConvert(Method evaluationMethod, Trigger trigger) {
+      try {
+        return (RunnerApi.Trigger) evaluationMethod.invoke(this, trigger);
```

Comment:
The casting code you have to write will be largely the same. When you pull a converter out of the map, it is the same as grabbing a method, and invoking requires the same sort of type-unsafe call.

The difference is really in writing a converter/visitor. It is very readable to write a bunch of overloads to cover the cases (somewhat like `DoFnSignature.Parameter.Cases`) but since anonymous functions are so bulky in Java 7 a map-based implementation adds a ton of readability overhead.

BTW the pattern here is loosely modeled on http://www.anzaan.com/2012/04/java-visitor-pattern-and-the-expression-problem-part-1-framework-based-approach/ and is also a modified version of the visitor logic from `TriggerStateMachines` (not shared because of dependency concerns). The reason for this approach is all about extensibility.

As it turns out, we currently don't have that many trigger syntax tree traversals, so I de-genericized the approach so it isn't really a reusable "TriggerEvaluator". That seemed like a bulky thing to put in the user-facing API. And perhaps we never will have more traversals, as we'll write them against the proto directly to make sure we cover all cases in the model, not just the Java SDK. I'm not sure.

My favorite solution as far as raw Java would be to make `Trigger` more like `DoFnSignature.Parameter` and then have its own `Cases`. Basically the same as a generic `TriggerEvaluator` but the terminology is a bit more obvious IMO. The backwards-compatibility breakage might start to be too much. We'd have to reconcile our seemingly open-ended fluent syntax for triggers with acknowledging that they are closed-world and have a centralized list of cases. I'm sure it is possible, just not taking it on right here.